### ISSUE TITLE
failure on `new WC_Order` stemming from empty `wc_get_order_types` after upgrade from 2.x.x to 3.1.x

### ISSUE THREAD

**Issue by fyaconiello** [woocommerce/woocommerce#16413]
## Prerequisites

- [X] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [X] The issue still exists against the latest `master` branch of WooCommerce on Github
- [X] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

1. upgrade from a 2.x version of woo commerce (i was upgrading from 2.6.14 to 3.1.1) to a 3.1.X
2. have any code, theme function or woocommerce addon plugin, that calls `new WC_Order($id);` with a valid order id
3. receive a fatal error `Invalid order.` thrown from `woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php`'s read function
3a. the logic that throws the exception has a check for `... || ! in_array( $post_object->post_type, wc_get_order_types() ) ...` and `wc_get_order_types()` always returns an empty array.

## Expected/actual behavior

Expected: a valid WC_Order object to get instantiated.
Actual: a fatal error.

## Isolating the problem

- [X] This bug happens with only WooCommerce plugin active
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [X] I can reproduce this bug consistently using the steps above

## WordPress Environment

<details>
```
`
### WordPress Environment ###

Home URL: https://dev-spicehunter.pantheonsite.io
Site URL: https://dev-spicehunter.pantheonsite.io
WC Version: 3.1.1
Log Directory Writable: ✔
WP Version: 4.8.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US

### Server Environment ###

Server Info: nginx/1.8.1
PHP Version: 7.0.18
PHP Post Max Size: 100 MB
PHP Time Limit: 120
PHP Max Input Vars: 10000
cURL Version: 7.40.0
NSS/3.21 Basic ECC

SUHOSIN Installed: –
Max Upload Size: 100 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.1.1
WC Database Prefix: wp_
woocommerce_sessions: ✔
woocommerce_api_keys: ✔
woocommerce_attribute_taxonomies: ✔
woocommerce_downloadable_product_permissions: ✔
woocommerce_order_items: ✔
woocommerce_order_itemmeta: ✔
woocommerce_tax_rates: ✔
woocommerce_tax_rate_locations: ✔
woocommerce_shipping_zones: ✔
woocommerce_shipping_zone_locations: ✔
woocommerce_shipping_zone_methods: ✔
woocommerce_payment_tokens: ✔
woocommerce_payment_tokenmeta: ✔
MaxMind GeoIP Database: ✔

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (26) ###

Gravity Forms: by rocketgenius – 2.2.3
Native PHP Sessions for WordPress: by Pantheon – 0.6.2
Advanced Custom Fields PRO: by Elliot Condon – 5.6.1
Better Search Replace: by Delicious Brains – 1.3
Capability Manager Enhanced: by Jordi Canals
Kevin Behrens – 1.5.7

Cloudflare: by John Wineman
Furkan Yilmaz
Junade Ali (Cloudflare Team) – 3.3.1

Gravity Forms Constant Contact Add-On: by Katz Web Services
Inc. – 2.2.2

MCE Table Buttons: by Jake Goldman
10up
Oomph – 3.2

PeopleSoft Order Export: by CO+LAB Multimedia
LLC – 1.0.0

Posts 2 Posts: by scribu – 1.6.5
Really Simple CSV Importer: by Takuro Hishikawa – 1.3
Regenerate Thumbnails: by Alex Mills (Viper007Bond) – 2.2.6
SearchWP: by Jonathan Christopher – 2.6.1
SendGrid: by SendGrid – 1.11.6
WooCommerce Print Invoice & Delivery Note: by Triggvy Gunderson – 4.1.5
WooCommerce Dynamic Pricing: by Lucas Stark – 3.0.12
WooCommerce Email Validation: by Hugh Lashbrooke – 2.1
WooCommerce Authorize.Net AIM Gateway: by SkyVerge – 3.12.0
WooCommerce Group Coupons: by itthinx – 1.7.1
WooCommerce Shop as Customer: by cxThemes – 2.10
WooCommerce Table Rate Shipping: by WooCommerce – 3.0.4
WooCommerce: by Automattic – 3.1.1
WordPress Importer: by wordpressdotorg – 0.6.3
Yoast SEO: by Team Yoast – 5.2
WP Migrate DB Pro Media Files: by Delicious Brains – 1.4.7
WP Migrate DB Pro: by Delicious Brains – 1.7.2

### Settings ###

API Enabled: ✔
Force SSL: ✔
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-1 (rated-1-2)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Shop base: #9107 - /products/filter/
Cart: #9082 - /cart/
Checkout: #9083 - /checkout/
My account: #9087 - /my-account/

### Theme ###

Name: The Spice Hunter
Version: 1.0.0
Author URL: http://www.teamcolab.com
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: spicehunter/woocommerce/archive-product.php
spicehunter/woocommerce/cart/cart-empty.php
spicehunter/woocommerce/cart/cart-totals.php
spicehunter/woocommerce/cart/cart.php
spicehunter/woocommerce/cart/cross-sells.php
spicehunter/woocommerce/checkout/form-billing.php
spicehunter/woocommerce/checkout/form-checkout.php
spicehunter/woocommerce/checkout/form-coupon.php
spicehunter/woocommerce/checkout/review-order.php
spicehunter/woocommerce/checkout/thankyou.php
spicehunter/woocommerce/content-product.php
spicehunter/woocommerce/content-single-product.php
spicehunter/woocommerce/emails/customer-on-hold-order.php
spicehunter/woocommerce/global/quantity-input.php
spicehunter/woocommerce/loop/loop-end.php
spicehunter/woocommerce/loop/loop-start.php
spicehunter/woocommerce/myaccount/my-orders.php
spicehunter/woocommerce/order/order-details-customer.php
spicehunter/woocommerce/order/order-details-item.php
spicehunter/woocommerce/order/order-details.php
spicehunter/woocommerce/single-product/add-to-cart/simple.php
spicehunter/woocommerce/single-product/meta.php
spicehunter/woocommerce/single-product/price.php
spicehunter/woocommerce/single-product/rating.php
spicehunter/woocommerce/single-product/related.php
spicehunter/woocommerce/single-product/review.php
spicehunter/woocommerce/single-product/tabs/tabs.php
spicehunter/woocommerce/single-product/title.php
spicehunter/woocommerce/single-product-reviews.php


### Authorize.Net AIM ###

Environment: Production
Debug Mode: Off
`
```
</details>

**Comment by mikejolley** [comment]
@fyaconiello how early is your call? Are you calling the above before `init`? If so, the types/post types will not be registered yet.

**Comment by mikejolley** [comment]
So basically, if you used the correct function (`wc_get_order`) to get the order instead of using classes, you'd see this notice:

https://github.com/woocommerce/woocommerce/blob/3522857d23051844bea742a37a1b0b242f7850af/includes/wc-order-functions.php#L81

> wc_get_order should not be called before post types are registered (woocommerce_after_register_post_type action)

Order types are registered at the same time as post types. Using orders before then will have consequences because certain data/functions/methods are not available yet.

tl;dr Make sure your custom code runs after `init`.
